### PR TITLE
Pyrometheus mixtures and chemistry

### PR DESCRIPTION
**Pull request by MTCam**
[Edit - new description]

This PR builds on (already-merged) scalar advection #95 and adds a _Pyrometheus_-based mixture EOS and chemistry source term interface.   It requires the MIRGE-compatible generator in this [_Pyrometheus_ PR](https://github.com/ecisneros8/pyrometheus/pull/7), and it requires _Cantera._

Several tests of both the generated _Pyrometheus_ code object, and the _Pyrometheus_-based EOS interface ensure that both agree very closely with _Cantera_.  Reactions are tested against a homogeneous reactor in Cantera. All tests are operated in the "MIRGE context" - i.e. with DOFArrays, and _MIRGE-Com_-specific data structures.

The `autoignition-mpi.py` example is added - which reproduces the autoignition verification case provided by Esteban.

### CODE REVIEW

**Comment by MTCam** [comment]
In this last push, I've switched to a natively implemented Prometheus mechanism because I perceived that I need to use the discretization and the array context *inside* the mechanism functions.   I hope we can talk about this, @inducer.

File: mirgecom/prometheus.py [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+
+        for iter in range(num_iter):
+            f = enthalpy_or_energy - he_fun(t_i, y)
+            j = -pv_fun(t_i, y)
+            dt = -f / j
+            t_i += dt
+            tresid = actx.np.abs(dt)
+            maxerr = self.discr.norm(tresid, np.inf)
+            if maxerr < tol:
+                break
```
Comment (MTCam) [inline mirgecom/prometheus.py:526]: This is the one-and-only place where I needed a norm, and had trouble figuring out how to compute one with DOFArrays sans discretization.

Comment (inducer) [reply mirgecom/prometheus.py:526]: OIC, makes sense (as discussed). Thanks!

**Comment by MTCam** [comment]
One thing I am not entirely happy about is the scattering about of CTI files in the MC directory structure.  Can we talk about what we should do with these (and potentially other) data files in general?

**Comment by MTCam** [comment]
Centralization of CTI mechanism files is here #207.

**Comment by MTCam** [comment]
#207 should merge here before review as it will clean up the addition of the CTI files.

File: examples/autoignition-mpi.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        cv = split_conserved(dim, q)
+        omega = eos.get_production_rates(cv)
+        w = eos.get_species_molecular_weights()
+        species_sources = w * omega
+        rho_source = 0 * cv.mass
+        mom_source = 0 * cv.momentum
+        energy_source = 0 * cv.energy
+        return join_conserved(dim, rho_source, energy_source, mom_source,
+                              species_sources)
+    sources = {my_chem_sources}
```
Comment (MTCam) [inline examples/autoignition-mpi.py:168]: Chemistry source terms calculated here.

File: examples/autoignition-mpi.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    def my_checkpoint(step, t, dt, state):
+        global checkpoint_t
+        cv = split_conserved(dim, state)
+        reaction_rates = eos.get_production_rates(cv)
+        viz_fields = [("reaction_rates", reaction_rates)]
+        return sim_checkpoint(discr, visualizer, eos, q=state,
+                              vizname=casename, step=step,
+                              t=t, dt=dt, nstatus=nstatus, nviz=nviz,
+                              constant_cfl=constant_cfl, comm=comm,
+                              viz_fields=viz_fields)
```
Comment (MTCam) [inline examples/autoignition-mpi.py:212]: This interface allows users to customize the fields that go into the viz files.  In this case stuffing `reaction_rates` into `viz_fields` does the trick.

**Comment by MTCam** [comment]
This stuff has been broken down into several more manageable pieces.   Please do not review here.

**Comment by MTCam** [comment]
Dead to me now.

**Pull request closed without merging**